File: src/main/java/at/searles/parsing/ParserStream.java
package at.searles.parsing;

import at.searles.buf.Frame;
import at.searles.lexer.TokenStream;
import at.searles.lexer.Tokenizer;
import at.searles.parsing.utils.ast.SourceInfo;

public class ParserStream {
    private final TokenStream stream;

    /**
     * Marks the start of the current parsed element.
     */
    private long parsedStart;

    /**
     * Marks the end of the current parsed element.
     */
    private long parsedEnd;
    private Listener listener;

    public ParserStream(TokenStream stream) {
        this.stream = stream;
        this.parsedStart = this.parsedEnd = stream.offset();
    }

    public TokenStream tokStream() {
        return stream;
    }

    public void setListener(Listener listener) {
        this.listener = listener;
    }

    public static ParserStream fromString(String string) {
        return new ParserStream(TokenStream.fromString(string));
    }

    public long getStart() {
        return this.parsedStart;
    }

    public long getEnd() {
        return this.parsedEnd;
    }

    public SourceInfo createSourceInfo() {
        return new RangeSourceInfo(this);
    }

    public void setStart(long start) {
        this.parsedStart = start;
    }

    public void setEnd(long end) {
        this.parsedEnd = end;
    }

    public CharSequence parseToken(Tokenizer tokenizer, int tokId, boolean exclusive) {
        Frame frame = tokenizer.matchToken(stream, tokId, exclusive);

        if(frame != null) {
            setStart(frame.startPosition());
            setEnd(frame.endPosition());
        }

        return frame;
    }

    /**
     * Sets the underlying stream to the given offset, ie,
     * the next token matched will start at the given parameter.
     * The caller must take care of start and end of the parsed
     * unit.
     *
     * @param offset The new offset.
     */
    public void backtrackToOffset(long offset) {
        stream.setPositionTo(offset);
    }

    /**
     * Returns the position from which the next token will be consumed
     */
    public long getOffset() {
        return stream.offset();
    }

    public String toString() {
        return stream.toString() + ": [" + parsedStart + ", " + parsedEnd + "]";
    }

    public <C> void notifyAnnotationBegin(C annotation) {
        if(listener != null) {
            listener.annotationBegin(this, annotation);
        }
    }

    /**
     * Override this method if necessary
     */
    public <C> void notifyAnnotationEnd(C annotation, boolean success) {
        if(listener != null) {
            listener.annotationEnd(this, annotation, success);
        }
    }

    private static class RangeSourceInfo implements SourceInfo {

        private final long end;
        private final long start;

        RangeSourceInfo(ParserStream stream) {
            this.start = stream.parsedStart;
            this.end = stream.parsedEnd;
        }

        @Override
        public long start() {
            return start;
        }

        @Override
        public long end() {
            return end;
        }

        public String toString() {
            return String.format("[%d:%d]", start(), end());
        }
    }

    public interface Listener {
        /**
         * If an annotation starts, the parser after it not necessarily succeeds
         * even in an LL1-Grammar. Yet, all calls to this method are
         * followed by a call to annotationEnd. Hence, all changes done
         * in this method must be undone if the arguments to annotationEnd
         * indicate that the annotation parser did not succeed.
         */
        <C> void annotationBegin(ParserStream parserStream, C annotation);

        <C> void annotationEnd(ParserStream parserStream, C annotation, boolean success);
    }

    public interface SimpleListener extends Listener {
        @Override
        default <C> void annotationBegin(ParserStream parserStream, C annotation) {}

        @Override
        default <C> void annotationEnd(ParserStream parserStream, C annotation, boolean success) {
            if(success) {
                annotate(parserStream, annotation);
            }
        }

        <C> void annotate(ParserStream parserStream, C annotation);
    }
}

File: src/main/java/at/searles/parsing/ParserStream.kt
package at.searles.parsing

import at.searles.buf.CharStream
import at.searles.buf.Frame
import at.searles.buf.ReaderCharStream
import at.searles.lexer.TokenStream
import at.searles.lexer.Tokenizer
import java.io.Reader

class ParserStream private constructor(private val tokenStream: TokenStream) {

    /**
     * Marks the start of the current parsed element.
     */
    var start: Long = tokenStream.offset
        private set

    /**
     * Marks the end of the current parsed element.
     */
    var end: Long = tokenStream.offset
        private set

    var maxStatus: BacktrackingStatus? = null
        private set

    var listener: Listener? = null

    /**
     * Returns the position from which the next token will be consumed
     */
    val offset: Long
        get() = tokenStream.offset


    /**
     * Sets the underlying stream to the given offset, ie,
     * the next token matched will start at the given parameter.
     * The caller must take care of start and end of the parsed
     * unit.
     *
     * @param offset The new offset.
     */
    fun restoreOffsetIfNecessary(offset: Long, failedParser: Any) {
        if(offset == this.offset) {
            // some empty parser failed. We are generous with them.
            return
        }

        if(maxStatus == null || maxStatus!!.requestedOffset < offset) {
            maxStatus = BacktrackingStatus(failedParser, offset, this)
        }

        tokenStream.offset = offset
    }

    fun createTrace(): Trace {
        return ParserStreamTrace(this)
    }

    fun parseToken(tokenizer: Tokenizer, tokenId: Int): Frame? {
        val frame = tokenizer.matchToken(tokenStream, tokenId)

        if (frame != null) {
            start = frame.start
            end = frame.end
        }

        return frame
    }

    fun recognize(recognizer: CanRecognize, isLeftMost: Boolean): Boolean {
        listener?.onTry(recognizer, this)

        val offset0 = offset
        val start0 = start
        val end0 = end

        if(recognizer.recognize(this)) {
            if(!isLeftMost) {
                start = start0
            }

            listener?.onSuccess(recognizer, this)

            return true
        }

        start = start0
        end = end0
        restoreOffsetIfNecessary(offset0, recognizer)

        listener?.onFail(recognizer, this)

        return false
    }

    fun <T> parse(parser: Parser<T>, isLeftMost: Boolean= true): T? {
        listener?.onTry(parser, this)

        val offset0 = offset
        val start0 = start
        val end0 = end

        val value = parser.parse(this)

        if(value != null) {
            if(!isLeftMost) {
                start = start0
            }

            listener?.onSuccess(parser, this)

            return value
        }

        start = start0
        end = end0
        restoreOffsetIfNecessary(offset0, parser)

        listener?.onFail(parser, this)

        return value
    }

    fun <T, U> reduce(left: T, reducer: Reducer<T, U>): U? {
        listener?.onTry(reducer, this)

        val offset0 = offset
        val start0 = start
        val end0 = end

        val value = reducer.reduce(left, this)

        start = start0 // position includes left

        if(value != null) {
            listener?.onSuccess(reducer, this)

            return value
        }

        end = end0
        restoreOffsetIfNecessary(offset0, reducer)

        listener?.onFail(reducer, this)

        return value
    }

    override fun toString(): String {
        return "$tokenStream: [$start, $end]"
    }

    fun notifyMark(marker: Any) {
        listener?.onMark(marker, this)
    }

    interface Listener {
        fun onMark(marker: Any, stream: ParserStream)
        fun onToken(tokenId: Int, frame: Frame, stream: ParserStream)
        fun onTry(parser: CanRecognize, stream: ParserStream)
        fun onSuccess(parser: CanRecognize, stream: ParserStream)
        fun onFail(parser: CanRecognize, stream: ParserStream)
    }

    class ParserStreamTrace(val stream: ParserStream) : Trace {
        override val start: Long = stream.start
        override val end: Long = stream.end

        override fun toString(): String {
            return "[$start : $end]"
        }
    }

    companion object {
        fun create(seq: CharSequence): ParserStream {
            return ParserStream(TokenStream.fromString(seq)).also {
                it.tokenStream.listener = object: TokenStream.Listener {
                    override fun onToken(tokenId: Int, frame: Frame, stream: TokenStream) {
                        it.listener?.onToken(tokenId, frame, it)
                    }
                }
            }
        }

        fun create(reader: Reader): ParserStream {
            return create(ReaderCharStream(reader))
        }

        fun create(stream: CharStream): ParserStream {
            return ParserStream(TokenStream.fromCharStream(stream)).also {
                it.tokenStream.listener = object: TokenStream.Listener {
                    override fun onToken(tokenId: Int, frame: Frame, stream: TokenStream) {
                        it.listener?.onToken(tokenId, frame, it)
                    }
                }
            }
        }
    }
}